orttraining/orttraining/python/training/ortmodule/_io.py

```diff
@@ -121,13 +147,9 @@ def flatten(self, args, kwargs, device):
         """Flatten args and kwargs in a single tuple of tensors with strict ordering"""
 
         ret = [_PrimitiveType.get_tensor(arg, device) if _PrimitiveType.is_primitive_type(arg) else arg for arg in args]
-        ret += [
-            _PrimitiveType.get_tensor(kwargs[name], device)
-            if _PrimitiveType.is_primitive_type(kwargs[name])
-            else kwargs[name]
-            for name in self.names
-            if name in kwargs
-        ]
+        flattened_kwargs = flatten_kwargs(kwargs, device)
+        ret += [flattened_kwargs[name] for name in self.names if name in flattened_kwargs]
```

Comment:
is there anywher used the [flattened_kwargs[name] for name in self.names if name in flattened_kwargs]?

Reply:
Looks with this change, the flatten result ret is only to make torch exporter happy, when running the FlattenModule forward, all kwargs are from self.kwargs instead of recovering from the flatten ones. If so, maybe we don't need flatten args also?

Reply:
torch onnx exporter is very picky about the **kwargs. It does not work well if the user model has **kwargs. So, to circumvent this problem, we had wrapped the original module inside another module which only accepts *args (not **kwargs) and relied on the fact that these new *args contain both the original *args as well as the original **kwargs (in a flattened way) so that the exporter can see all the tensors that need to be used in onnx graph building. So, flattening is needed (I believe). But let me evaluate it on a model to see if we can probably get rid of the flattening.

Reply:
Tried your suggestion locally and it failed. I think we need to flatten all tensors so the tracer can see the inputs.

Reply:
I am curious why it fails, because we did not do this for args?

_PrimitiveType.get_tensor(arg, device) if _PrimitiveType.is_primitive_type(arg)

Reply:
I am actually not sure where the error comes from. The error I see is as follows:

```py
E           Traceback (most recent call last):
E             File "/home/bmeswani/github/onnxruntime/build/Linux/RelWithDebInfo/onnxruntime/training/ortmodule/_utils.py", line 318, in get_exception_as_string
E               raise exception
E             File "/home/bmeswani/github/onnxruntime/build/Linux/RelWithDebInfo/onnxruntime/training/ortmodule/_graph_execution_manager.py", line 408, in _get_exported_model
E               torch.onnx.export(
E             File "/bert_ort/bmeswani/py39/lib/python3.9/site-packages/torch/onnx/__init__.py", line 350, in export
E               return utils.export(
E             File "/bert_ort/bmeswani/py39/lib/python3.9/site-packages/torch/onnx/utils.py", line 163, in export
E               _export(
E             File "/bert_ort/bmeswani/py39/lib/python3.9/site-packages/torch/onnx/utils.py", line 1074, in _export
E               graph, params_dict, torch_out = _model_to_graph(
E             File "/bert_ort/bmeswani/py39/lib/python3.9/site-packages/torch/onnx/utils.py", line 731, in _model_to_graph
E               graph = _optimize_graph(
E             File "/bert_ort/bmeswani/py39/lib/python3.9/site-packages/torch/onnx/utils.py", line 306, in _optimize_graph
E               _C._jit_pass_onnx_set_dynamic_input_shape(graph, dynamic_axes, input_names)
E           RuntimeError: i < graph->inputs().size() INTERNAL ASSERT FAILED at "../torch/csrc/jit/passes/onnx/shape_type_inference.cpp":2077, please report a bug to PyTorch.
```


The exporter must be able to know the inputs to the model when we pass the `sample_inputs_as_tuple` to the exporter. If we do not flatten, the `sample_inputs_as_tuple` does not contain these kwargs tensors and the exporter does not understand the inputs. This is just my guess; I would probably need someone from the exporter team to answer why this fails.